File: app/src/main/java/com/jds/fitnessjunkiess/getfitapp/features/exercisesDatabase/ExercisesListFragment.java
package com.jds.fitnessjunkiess.getfitapp.features.exercisesDatabase;

import android.content.Context;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import com.jds.fitnessjunkiess.getfitapp.R;
import com.jds.fitnessjunkiess.getfitapp.data.dataModels.Exercise;
import com.jds.fitnessjunkiess.getfitapp.data.dataModels.Workout;
import com.jds.fitnessjunkiess.getfitapp.data.dataModels.WorkoutExerciseAssignment;
import com.jds.fitnessjunkiess.getfitapp.features.exercisesDatabase.adapters.AbstractExercisesAdapter;
import com.jds.fitnessjunkiess.getfitapp.features.exercisesDatabase.adapters.ExercisesAdapter;
import com.jds.fitnessjunkiess.getfitapp.features.exercisesDatabase.adapters.WorkoutContextExercisesAdapter;

import java.util.List;

public class ExercisesListFragment extends Fragment implements AbstractExercisesAdapter.OnItemMenuClickInterface {
  private OnFragmentInteractionListener mListener;
  private AbstractExercisesAdapter recyclerViewerAdapter;

  public ExercisesListFragment() {
    // Required empty public constructor
  }

  @Override
  public void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
  }

  @Override
  public View onCreateView(LayoutInflater inflater, ViewGroup container,
                           Bundle savedInstanceState) {

    View rootView = inflater.inflate(R.layout.fragment_exercises_list, container, false);

    RecyclerView recyclerView = rootView.findViewById(R.id.exercise_view_recycle_viewer);
    RecyclerView.LayoutManager layoutManager = new LinearLayoutManager(rootView.getContext());
    recyclerView.setLayoutManager(layoutManager);

    if (this.mListener.getActiveWorkout() == null) {
      this.recyclerViewerAdapter = new ExercisesAdapter(this);
    } else {
      this.recyclerViewerAdapter = new WorkoutContextExercisesAdapter(this);
    }

    recyclerView.setAdapter(recyclerViewerAdapter);

    return rootView;
  }

  @Override
  public void onAttach(Context context) {
    super.onAttach(context);
    if (context instanceof OnFragmentInteractionListener) {
      mListener = (OnFragmentInteractionListener) context;
    } else {
      throw new RuntimeException(context.toString()
          + " must implement OnFragmentInteractionListener");
    }
  }

  @Override
  public void onDetach() {
    super.onDetach();
    mListener = null;
  }

  public void updateDataSet(List<Exercise> exercises) {
    this.recyclerViewerAdapter.updateDataSet(exercises);
  }

  @Override
  public Context getAppContext() {
    return getContext();
  }

  @Override
  public List<Workout> getWorkoutsList() {
    return this.mListener.getWorkoutsList();
  }

  @Override
  public void insertExerciseAssignment(Exercise exercise, Workout workout) {
    WorkoutExerciseAssignment exerciseAssignment = new WorkoutExerciseAssignment();
    exerciseAssignment.setExerciseId(exercise.getId());
    exerciseAssignment.setWorkoutId(workout.getId());
    this.mListener.insertExerciseAssignment(exerciseAssignment);

    Toast.makeText(
        getContext(),
        exercise.getName() + " added to " + workout.getName(),
        Toast.LENGTH_LONG).show();
  }

  @Override
  public void insertExerciseAssignment(Exercise exercise) {
    WorkoutExerciseAssignment exerciseAssignment = new WorkoutExerciseAssignment();
    exerciseAssignment.setExerciseId(exercise.getId());
    exerciseAssignment.setWorkoutId(this.mListener.getActiveWorkout().getId());
    this.mListener.insertExerciseAssignment(exerciseAssignment);

    Toast.makeText(
        getContext(),
        exercise.getName() + " added to " + this.mListener.getActiveWorkout().getName(),
        Toast.LENGTH_LONG).show();
  }

  public interface OnFragmentInteractionListener {
    Workout getActiveWorkout();
    List<Workout> getWorkoutsList();
    void insertExerciseAssignment(WorkoutExerciseAssignment workoutExerciseAssignment);
  }
}

File: app/src/main/java/com/jds/fitnessjunkiess/getfitapp/features/exercisesDatabase/ExercisesListFragment.kt
package com.jds.fitnessjunkiess.getfitapp.features.exercisesDatabase

import android.content.Context
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import com.jds.fitnessjunkiess.getfitapp.R
import com.jds.fitnessjunkiess.getfitapp.data.dataModels.Exercise
import com.jds.fitnessjunkiess.getfitapp.data.dataModels.Workout
import com.jds.fitnessjunkiess.getfitapp.data.dataModels.WorkoutExerciseAssignment
import com.jds.fitnessjunkiess.getfitapp.features.exercisesDatabase.adapters.AbstractExercisesAdapter
import com.jds.fitnessjunkiess.getfitapp.features.exercisesDatabase.adapters.ExercisesAdapter
import com.jds.fitnessjunkiess.getfitapp.features.exercisesDatabase.adapters.WorkoutContextExercisesAdapter

class ExercisesListFragment : Fragment(),
    AbstractExercisesAdapter.OnItemMenuClickInterface {

    private var mListener: OnFragmentInteractionListener? = null
    private lateinit var recyclerViewerAdapter: AbstractExercisesAdapter

    override val workoutsList: List<Workout>
        get() = mListener?.workoutsList!!

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        super.onCreateView(inflater, container, savedInstanceState)

        val rootView: View = inflater.inflate(
            R.layout.fragment_exercises_list, container, false
        )
        val recyclerView = rootView.findViewById<RecyclerView>(
            R.id.exercise_view_recycle_viewer
        )

        val layoutManager = LinearLayoutManager(rootView.context)
        recyclerView.layoutManager = layoutManager

        if (this.mListener?.activeWorkout == null) {
            this.recyclerViewerAdapter = ExercisesAdapter(context, this)
        } else {
            this.recyclerViewerAdapter = WorkoutContextExercisesAdapter(this)
        }

        recyclerView.adapter = recyclerViewerAdapter

        return rootView
    }

    override fun onAttach(context: Context?) {
        super.onAttach(context)
        if (context is OnFragmentInteractionListener) {
            mListener = context
        } else {
            throw RuntimeException(
                "${context!!.toString()} must implement OnFragmentInteractionListener"
            )
        }
    }

    override fun onDetach() {
        super.onDetach()
        mListener = null
    }

    fun updateDataSet(exercises: List<Exercise>) {
        this.recyclerViewerAdapter.updateDataSet(exercises)
    }

    override fun insertExerciseAssignment(exercise: Exercise, workout: Workout) {
        val exerciseAssignment = WorkoutExerciseAssignment()
        exerciseAssignment.exerciseId = exercise.id
        exerciseAssignment.workoutId = workout.id

        this.mListener?.insertExerciseAssignment(exerciseAssignment)

        Toast.makeText(
            context,
            "${exercise.name} added to ${workout?.name}",
            Toast.LENGTH_LONG
        ).show()
    }

    override fun insertExerciseAssignment(exercise: Exercise) {
        val exerciseAssignment = WorkoutExerciseAssignment()
        exerciseAssignment.exerciseId = exercise.id
        exerciseAssignment.workoutId = this.mListener!!.activeWorkout.id
        this.mListener!!.insertExerciseAssignment(exerciseAssignment)

        Toast.makeText(
            context,
            "${exercise.name} added to ${this.mListener!!.activeWorkout.name}",
            Toast.LENGTH_LONG
        ).show()
    }

    interface OnFragmentInteractionListener {
        val activeWorkout: Workout
        val workoutsList: List<Workout>
        fun insertExerciseAssignment(workoutExerciseAssignment: WorkoutExerciseAssignment)
    }
}